### ISSUE TITLE
[Chore]: Connection error in the E2E tests

### ISSUE THREAD

**Issue by georgethebeatle** [cloudfoundry/korifi#1310]
### Build URL

https://ci.korifi.cf-app.com/teams/main/pipelines/main/jobs/run-e2es-periodic/builds/363
https://ci.korifi.cf-app.com/teams/main/pipelines/main/jobs/run-e2es-periodic/builds/349
https://ci.korifi.cf-app.com/teams/main/pipelines/main/jobs/run-e2es-periodic/builds/324
https://ci.korifi.cf-app.com/teams/main/pipelines/main/jobs/run-e2es-periodic/builds/338
https://ci.korifi.cf-app.com/teams/main/pipelines/main/jobs/run-e2es-periodic/builds/323
https://ci.korifi.cf-app.com/teams/main/pipelines/main/jobs/run-e2es-periodic/builds/280
https://ci.korifi.cf-app.com/teams/main/pipelines/main/jobs/run-e2es-pr/builds/173
https://ci.korifi.cf-app.com/teams/main/pipelines/main/jobs/run-e2es-periodic/builds/211
https://ci.korifi.cf-app.com/teams/main/pipelines/main/jobs/run-e2es-periodic/builds/138
https://ci.korifi.cf-app.com/teams/main/pipelines/main/jobs/run-e2es-pr/builds/43.1
https://ci.korifi.cf-app.com/teams/main/pipelines/main/jobs/run-e2es-pr/builds/43

### Description

Different tests in the E2E suite seem to be failing sporadically with connection errors. See log bellow for an example failure.

```
• [FAILED] [4.213 seconds]
Orgs
/tmp/build/3e139e9d/korifi/tests/e2e/orgs_test.go:16
  list domains
  /tmp/build/3e139e9d/korifi/tests/e2e/orgs_test.go:277
    when the user is not authorized in the organization
    /tmp/build/3e139e9d/korifi/tests/e2e/orgs_test.go:313
      [It] returns a not found error
      /tmp/build/3e139e9d/korifi/tests/e2e/orgs_test.go:318

  Begin Captured StdOut/StdErr Output >>
    W0703 18:26:11.358444     519 gcp.go:120] WARNING: the gcp auth plugin is deprecated in v1.22+, unavailable in v1.25+; use gcloud instead.
    To learn more, consult https://cloud.google.com/blog/products/containers-kubernetes/kubectl-auth-changes-in-gke
  << End Captured StdOut/StdErr Output

  Begin Captured GinkgoWriter Output >>
    Fail Handler: failure correlation ID: "e91a4d07-428e-4dcc-a512-c65b9d1f9a4f"


    ===== Logs for pod "korifi-api-deployment-ccbb6796d-brhjs" with correlation ID "e91a4d07-428e-4dcc-a512-c65b9d1f9a4f" (last 50 lines) =====
    {"level":"info","ts":1656872768.52545,"logger":"retrying-client","msg":"k8s client returned forbidden","correlation-id":"e91a4d07-428e-4dcc-a512-c65b9d1f9a4f","op":"create","count":1,"error":"rolebindings.rbac.authorization.k8s.io \"cf-98ffe9c334e24367195b34585ce035a0c368a8f9762a6ade1d2ed5c8975f7afd\" already exists"}

    ==============================================



    ===== Logs for pod "korifi-controllers-controller-manager-6d89b879f-pfhrl" (last 50 lines) =====
    1.656872771118818e+09	DEBUG	controller-runtime.webhook.webhooks	wrote response	{"webhook": "/validate-korifi-cloudfoundry-org-v1alpha1-cfapp", "code": 200, "reason": "", "UID": "6667c493-55e9-40c4-854d-067bb947a504", "allowed": true}
    1.6568727714641182e+09	DEBUG	controller-runtime.webhook.webhooks	received request	{"webhook": "/mutate-korifi-cloudfoundry-org-v1alpha1-cfroute", "UID": "4c3c0d93-c6bb-4842-a536-0392497ec714", "kind": "korifi.cloudfoundry.org/v1alpha1, Kind=CFRoute", "resource": {"group":"korifi.cloudfoundry.org","version":"v1alpha1","resource":"cfroutes"}}
    1.656872771464388e+09	INFO	cfroute-resource	Mutating CFRoute webhook handler	{"name": "cf-route-ce9a2308-e12c-43d6-8843-7f2ef37e9d40"}
    1.656872771464774e+09	DEBUG	controller-runtime.webhook.webhooks	wrote response	{"webhook": "/mutate-korifi-cloudfoundry-org-v1alpha1-cfroute", "code": 200, "reason": "", "UID": "4c3c0d93-c6bb-4842-a536-0392497ec714", "allowed": true}
    1.6568727714651327e+09	DEBUG	controller-runtime.webhook.webhooks	received request	{"webhook": "/validate-korifi-cloudfoundry-org-v1alpha1-cfspace", "UID": "9de8e0f5-7f9a-41f4-b10f-12225cfbcbee", "kind": "korifi.cloudfoundry.org/v1alpha1, Kind=CFSpace", "resource": {"group":"korifi.cloudfoundry.org","version":"v1alpha1","resource":"cfspaces"}}
    1.6568727714787273e+09	DEBUG	controller-runtime.webhook.webhooks	received request	{"webhook": "/validate-korifi-cloudfoundry-org-v1alpha1-cfroute", "UID": "0fd113e7-1c8e-45cd-a199-2ace619f8966", "kind": "korifi.cloudfoundry.org/v1alpha1, Kind=CFRoute", "resource": {"group":"korifi.cloudfoundry.org","version":"v1alpha1","resource":"cfroutes"}}
    1.6568727714791434e+09	ERROR	route-validation	Route destination app not found in space	{"error": "CFApp.korifi.cloudfoundry.org \"18887ed9-5cc8-48b8-b48e-cec161478f98\" not found"}
    code.cloudfoundry.org/korifi/controllers/webhooks/networking.(*CFRouteValidator).validateDestinations
    	/workspace/controllers/webhooks/networking/cfroute_validator.go:207
    code.cloudfoundry.org/korifi/controllers/webhooks/networking.(*CFRouteValidator).ValidateUpdate
    	/workspace/controllers/webhooks/networking/cfroute_validator.go:128
    sigs.k8s.io/controller-runtime/pkg/webhook/admission.(*validatorForType).Handle
    	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.12.2/pkg/webhook/admission/validator_custom.go:87
    sigs.k8s.io/controller-runtime/pkg/webhook/admission.(*Webhook).Handle
    	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.12.2/pkg/webhook/admission/webhook.go:146
    sigs.k8s.io/controller-runtime/pkg/webhook/admission.(*Webhook).ServeHTTP
    	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.12.2/pkg/webhook/admission/http.go:98
    github.com/prometheus/client_golang/prometheus/promhttp.InstrumentHandlerInFlight.func1
    	/go/pkg/mod/github.com/prometheus/client_golang@v1.12.2/prometheus/promhttp/instrument_server.go:40
    net/http.HandlerFunc.ServeHTTP
    	/usr/local/go/src/net/http/server.go:2084
    github.com/prometheus/client_golang/prometheus/promhttp.InstrumentHandlerCounter.func1
    	/go/pkg/mod/github.com/prometheus/client_golang@v1.12.2/prometheus/promhttp/instrument_server.go:117
    net/http.HandlerFunc.ServeHTTP
    	/usr/local/go/src/net/http/server.go:2084
    github.com/prometheus/client_golang/prometheus/promhttp.InstrumentHandlerDuration.func2
    	/go/pkg/mod/github.com/prometheus/client_golang@v1.12.2/prometheus/promhttp/instrument_server.go:84
    net/http.HandlerFunc.ServeHTTP
    	/usr/local/go/src/net/http/server.go:2084
    net/http.(*ServeMux).ServeHTTP
    	/usr/local/go/src/net/http/server.go:2462
    net/http.serverHandler.ServeHTTP
    	/usr/local/go/src/net/http/server.go:2916
    net/http.(*conn).serve
    	/usr/local/go/src/net/http/server.go:1966
    1.656872771479265e+09	DEBUG	controller-runtime.webhook.webhooks	wrote response	{"webhook": "/validate-korifi-cloudfoundry-org-v1alpha1-cfroute", "code": 403, "reason": "{\"validationErrorType\":\"RouteDestinationNotInSpaceError\",\"message\":\"Route destination app not found in space\"}", "UID": "0fd113e7-1c8e-45cd-a199-2ace619f8966", "allowed": false}
    1.656872771488808e+09	DEBUG	controller-runtime.webhook.webhooks	wrote response	{"webhook": "/validate-korifi-cloudfoundry-org-v1alpha1-cfspace", "code": 200, "reason": "", "UID": "9de8e0f5-7f9a-41f4-b10f-12225cfbcbee", "allowed": true}
    1.6568727715045846e+09	INFO	controllers.CFSpace	Reconciling deletion of cf-space-0c2489b4-ee43-4521-bf35-ef199ab86e11
    1.6568727715612676e+09	DEBUG	controller-runtime.webhook.webhooks	received request	{"webhook": "/validate-korifi-cloudfoundry-org-v1alpha1-cfspace", "UID": "22b1b202-a4ff-4ada-8023-f35579973ff0", "kind": "korifi.cloudfoundry.org/v1alpha1, Kind=CFSpace", "resource": {"group":"korifi.cloudfoundry.org","version":"v1alpha1","resource":"cfspaces"}}
    1.6568727715616474e+09	DEBUG	controller-runtime.webhook.webhooks	wrote response	{"webhook": "/validate-korifi-cloudfoundry-org-v1alpha1-cfspace", "code": 200, "reason": "", "UID": "22b1b202-a4ff-4ada-8023-f35579973ff0", "allowed": true}
    1.6568727716107204e+09	ERROR	controllers.CFSpace	Error when trying to fetch CFSpace cf-org-8d0f45bb-1437-463d-80cb-9b307f11f1d1/cf-space-0c2489b4-ee43-4521-bf35-ef199ab86e11	{"error": "CFSpace.korifi.cloudfoundry.org \"cf-space-0c2489b4-ee43-4521-bf35-ef199ab86e11\" not found"}
    sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Reconcile
    	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.12.2/pkg/internal/controller/controller.go:121
    sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler
    	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.12.2/pkg/internal/controller/controller.go:320
    sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem
    	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.12.2/pkg/internal/controller/controller.go:273
    sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2
    	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.12.2/pkg/internal/controller/controller.go:234
    1.6568727716326892e+09	DEBUG	controller-runtime.webhook.webhooks	received request	{"webhook": "/validate-korifi-cloudfoundry-org-v1alpha1-cfspace", "UID": "5aa9204f-c2ed-4a82-9890-0bd9122abc2e", "kind": "korifi.cloudfoundry.org/v1alpha1, Kind=CFSpace", "resource": {"group":"korifi.cloudfoundry.org","version":"v1alpha1","resource":"cfspaces"}}

    ==============================================

    Fail Handler: completed
  << End Captured GinkgoWriter Output

  Expected
      <helpers.actualRestyResponse>: 
      Request: GET https://cf.pr-e2e.korifi.cf-app.com/v3/organizations/cf-org-9f546b5b-e0cd-47f2-9074-9ae537efb8d3/domains
      Request headers:
          X-Correlation-Id: e91a4d07-428e-4dcc-a512-c65b9d1f9a4f
          User-Agent: go-resty/2.7.0 (https://github.com/go-resty/resty)
          Authorization: Bearer <Redacted>
      Request body:
          null
      HTTP Status code: 503
      Response headers:
          Content-Type: text/plain
          Vary: Accept-Encoding
          Date: Sun, 03 Jul 2022 18:26:11 GMT
          Server: envoy
      Response body:
          upstream connect error or disconnect/reset before headers. reset reason: connection failure
  to have HTTP Status code
      <int>: 404
  In [It] at: /go/pkg/mod/github.com/onsi/gomega@v1.19.0/internal/assertion.go:58
```

**Comment by georgethebeatle** [comment]
Looking at the failure times and superimposing them on the dashboard data for both pr-e2e and concourse clusters it looks like the connection reset is not due to particularly high load or because test run coincided with cluster recreation. 

Unfortunately there are not too many logs from the api shim (they are all filtered by correlation id) so not too much insight if the shim crashed.

Could it be some default server configuration that is throttling requests if they get too many?

**Comment by danail-branekov** [comment]
There are some evidences on the internet that a `connection reset` could be caused by a timeout between the contour proxy and the api server (i.e. the api server is too slow handling the request). However, in the korifi deployment we set this timeout [to 5 minutes](https://github.com/cloudfoundry/korifi/blob/ad0ff5031568177679ea3940091e5fb8469a99ed/api/config/base/ingress.yaml#L22), therefore it is not likely to be that. 

I have added an "on-failure" task to e2e test (both main and periodics) to [get contour's logs](https://github.com/cloudfoundry/korifi-ci/commit/2ab53af06cad1e84f74e2a0dd1089e39503d610d), hopefully this should shed more light on the issue.

Blocking until the issue happens again

**Comment by danail-branekov** [comment]
We have it reproduced on CI with contour logs: https://ci.korifi.cf-app.com/teams/main/pipelines/main/jobs/run-e2es-main/builds/104

**Comment by julian-hj** [comment]
I believe that this was caused by panics in the API shim. We found some evidence in the logs that some logger calls were passing non-even arguments in the key/value part of the argument list, which causes a panic. That would account for Contour losing connection to the API.

We think this *should* be fixed by [this commit](https://github.com/cloudfoundry/korifi/commit/f67030321f20a5a2b5fdaa38dab79e1bc1e59c50)

Separately, on a procedural note, should this chore not be a bug? Having the API report 503s feels like a bug to me.
